README.md
# redoubt 🏰

redoubt is a simple, yet powerful & secure [express](https://github.com/expressjs/express/) based server.

## Why?

I recently came to realize that most of my projects incorporating a web server end up containing the exact same server boilerplate code:

 * A JSON Body Parser
 * A URL Encoded Parser
 * A static file directory somewhere
 * An (optional) cookie based session storage
 * Proper compression for all files & data
 * And last but not least SSL certificates

The first four are rather trivial to set up, the last one would travidionally be a bit tricky and/or costly. 

Thankfully, [Let's Encrypt](https://letsencrypt.org/) is giving aways certificates, so we've got **free, automatically renewed certificates** built in.

## Quick Start

Install the library into your project:

```
npm i redoubt
```

You are ready to set up your server:

```js
const Redoubt = require("redoubt");
const path = require("path");
const fs = require("fs");

const server = new Redoubt({
  // The name of your server
  name: "my-redoubt-server",
  // Let's Encrypt will issue our certificates(We need to specify some details about our domain)
  certs: "letsEncrypt",
  agreeGreenlockTos: true,
  domains: ["patrick-sachs.de"],
  webmasterMail: "the-webmaster-mail@patrick-sachs.de"
  // We also serve some static files
  staticFiles: { from: path.resolve(__dirname, "./static"), serve: "/" },
  // The cookie on the client side(contains the session ID) is obfuscated with this secret.
  cookieSecret: "maggots-at-the-party",
});
// Express app exposed via "app" - Set up your actual server here!
server.app.get("/api/data", (req, res) => res
  .json({ values: ["express", "test"] })
  .end());
// Let's rock! 🤘
server.listen();
```

The code above is an example for a production server. Production servers using `certs: "letsEncrypt"` need a public FQDN in order for Let's Encrypt to be able to issue certificates to it.

For development server, we can either set `isDevelopment: true`, or use manually created certificates:

```js
certs: {
  cert: fs.readFileSync(path.resolve(__dirname, "./server.crt")),
  key: fs.readFileSync(path.resolve(__dirname, "./key.pem")),
  allowUnsigned: true
},
```

See the `/example/index.js` file for a complete development server example.

## All properties

| Key | Description | Type | Default/Required |
| - | - | - | - |
| `name` | The name of your server. Currently only used for the name of the session cookie, but this might change in the future. | `string` | ✔️ Required |
| `domains` | The domain names the Let's Encrypt certificates should be issued against. The first value is your primary domain. Must all be valid, resolvable FQDNs. Simply pass `localhost` in a development envrionment. | `string[]` | ✔️ Required |
| `webmasterMail` | The mail of the webmaster. Used by Let's Encrypt to conact you when something related to your certificates happens(about to be renewed...). | `string` | ✔️ Required |
| `cookieSecret` | This secret is used to obfuscate the session ID of the user saved in the cookie. Only the ID of the session is saved on the client side, not the actual data. If this secret is weak or compromised in any other way users can trivially pose as other users and access their session data. | `string` | ✔️ Required |
| `isDevelopment` | If this server running in development mode? Development mode servers do not compress, do not request valid SSL certificates and trust invalid SSL certificates. | `boolean` | `false` |
| `ssl` | The certificates your server uses. You can either use Let's Encrypt or your your own certificates(e.g. if you are running in a company internal network). Keep in mind that when using manually generated certificates, `key` and `cert` must contain the certificate & key file contents, not their file paths. | `"none" | "letsEncrypt" | { key: string, cert: string, allowUnsigned: boolean }` | `"letsEncrypt"` |
| `letsEncryptCertDirectory` | When using Let's Encrypt certificates, this is where they will be stored. Must be a directory with write access. | `string` | `"./.certs/"` |
| `agreeGreenlockTos` | The greenlock library used to create the Let's Encrypt certificates requires you to explictly accept their TOS. | `boolean` | `false` |
| `staticFiles` | Allows you to serve static files. File location on your server is specified in `from`, and URL users will be able to access them from in `serve`. | `{ from: string, serve: string } | null` | `null` |
| `maxPayloadSize` | The max size in bytes a single JSON or URL encoded request to the server can have. | `number` | `100 * 1024` |
| `debug` | Allows to to specify your own debug function(Or disable debug by passing `null`). By default the console is used. | `((level: string, ...args: any[]) => void) | null` | `console.log("[" + level + "]", "🏰 ", ...args)` |

## Questions

 * *I'm getting an error about greenlock TOS!* - Please read about the `agreeGreenlockTos` option.
 * *My development server is "Not Secure"!* - This is expected. Let's Encrypt can't issue certificates for `localhost`. As long as the `domains` option if correct on your production server it will work there.
 * *I don't want to use port 443 & 80!* - The `listen(https: number, http: number | null)` function allows you to customize the ports. You can also pass null as second parameter to disable the HTTP server redirection.
 * *How do I access the user session data?* - See the express-session documentation: https://www.npmjs.com/package/express-session

## Contributing

Feel free to go ahead and do so. I'm very open when it comesto PRs, Issues and Feature Requests.

This project is written in TypeScript and has a very minimal setup and configuration.


CHANGELOG.md
# Changelog redoubt

All notable changes to this library will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this library adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## 2.0.2-rc2

- Add the `ca` parameter to add CAs for custom certificates.

## 2.0.2-rc1

- Add the `host` parameter to the `listen` function to optimally bind to a host/IP. Beta feature.

## 2.0.1

- Fixed an invalid setting sfor non secure cookies.

## 2.0.0

* Rename `certs` option to `ssl` and add a `none` option to disable HTTPs.
* Add a `close` function to shut the server down.
* Expose all created servers in the `server` property.

## 1.1.0

* Update dependencies

## 1.0.2

* Fixed import statements to better work with TypeScript @PatrickSachs.

## 1.0.1

* Fixed some typos @PatrickSachs.
* Changed TypeScript to be a devDependency @PatrickSachs.

## 1.0.0

* Initial commit @PatrickSachs.

src/index.ts
import express from "express";
import spdy from "spdy";
import greenlock from "greenlock-express";
import compression from "compression";
import { EventEmitter } from "events";
import session from "express-session";
import net from "net";
import http from "http";
import https from "https";

type DestroyFunction = (cb?: Function) => HttpOrHttpsServer;
type HttpOrHttpsServer = (http.Server | https.Server) & { destroy?: DestroyFunction };

// enableDestroy(server) is based on https://github.com/isaacs/server-destroy
const enableDestroy = (server0: http.Server | https.Server) => {
  const server: HttpOrHttpsServer = server0 as HttpOrHttpsServer;
  const connections: { [key: string]: net.Socket } = {}

  server.on('connection', function (conn: net.Socket) {
    const key = conn.remoteAddress + ':' + conn.remotePort;
    connections[key] = conn;
    conn.once('close', function () {
      delete connections[key];
    });
  });

  server.destroy = function (cb) {
    server.close(cb);
    for (let key in connections) {
      connections[key].destroy();
    }
    return server;
  };

  return server;
}

class Server extends EventEmitter {
  constructor(options: Partial<IOptions>) {
    super();
    this._options = createDefaultOptions(options);
    this._options.debug && this._options.debug("debug", "Options", this._options);
    this._app = express();
    this.basicSetup();
  }

  private _options: IOptions;
  private _app: express.Express;
  private _server: HttpOrHttpsServer[] = [];
  private _portHttp: number | null = null;
  private _portHttps: number | null = null;
  private _host: string | null = null;

  // ========================================
  // PUBLIC API
  // ========================================

  // Expose express

  public get app() {
    return this._app;
  }

  public get server() {
    return this._server;
  }

  // Lifecycle

  /**
   * Gracefully closes this redoubt server.
   * @param now Should the server be closed now(terminates all active connections) or wait for all active connections to terminate and then shut down?
   */
  public async close(now = false) {
    this._portHttp = null;
    this._portHttps = null;
    const promises: Promise<HttpOrHttpsServer>[] = [];
    for (let server of this._server) {
      const srv = server;
      promises.push(new Promise((resolve, reject) => {
        const callback = (err: any) => err ? reject(err) : resolve(srv);
        if (now && srv.destroy) {
          srv.destroy(callback);
        } else {
          srv.close(callback)
        };
      }));
    }
    this._server = [];
    return await Promise.all(promises);
  }

  /**
   * Starts the server & listens on the given ports.
   * @param https The HTTPS port to run your server on.
   * @param http A HTTP port to also listen to. Simply redirecty to HTTPS. Set to `null` to disable.
   */
  public async listen(https = 443, http: number | null = 80, host: string | null = null) {
    this._portHttp = http;
    this._portHttps = https;
    this._host = host;

    return await this.startListening();
  }

  // ========================================
  // PRIVATE FUNCTIONS
  // ========================================

  private emitEvent(event: string) {
    this.emit(event, this);
  }

  private basicSetup() {
    this._app.use(compression({
      filter: (req, res) => {
        if (this._options.isDevelopment) return false;
        if (req.headers["x-no-compression"]) return false;
        return compression.filter(req, res);
      }
    }));
    this._app.use(express.json({ limit: this._options.maxPayloadSize }));
    this._app.use(express.urlencoded({ limit: this._options.maxPayloadSize, extended: true }));
    this._options.staticFiles && this._app.use(this._options.staticFiles.serve, express.static(this._options.staticFiles.from));
    this._app.use(session({
      name: this._options.name,
      cookie: {
        // Prevent reading the session cookie from script
        httpOnly: true,
        // The cookie is only valid for our site.
        sameSite: true,
        // Don't send cookie over HTTP.
        // Use server.app.set("trust proxy", 1); when using a proxy and having problems with cookies.
        secure: this._options.ssl !== "none"
      },
      resave: false,
      // Don't save cookies if nothing has been saved in the session to comply with stricter cookie regulations.
      saveUninitialized: false,
      unset: "destroy",
      secret: this._options.cookieSecret
    }));
  }


  private greenlockOptions() {
    return {
      // It is not allowed to build apps/libraries which do not require the user to explictly accept the greenlock TOS.
      agreeTos: this._options.agreeGreenlockTos,
      version: "draft-11",
      // If development mode contact the staging server. They issue invalid certficates.
      server: this._options.isDevelopment ? "https://acme-staging-v02.api.letsencrypt.org/directory" : "https://acme-v02.api.letsencrypt.org/directory",
      // Let's encrypt requires a valid mail to contact you whenever something is going on with your certs.
      email: this._options.webmasterMail,
      // The domains the certs belong to.
      approveDomains: this._options.domains,
      // Where to save the certs.
      configDir: this._options.letsEncryptCertDirectory,
      // Debug on?
      debug: !!this._options.debug,
      // Sorry, but we'll opt out.
      communityMember: false,
      telemetry: false
    }
  };

  private spdyOptions() {
    return {
      key: typeof this._options.ssl !== "string" ? this._options.ssl.key : "",
      cert: typeof this._options.ssl !== "string" ? this._options.ssl.cert : "",
      ca: typeof this._options.ssl !== "string" ? (this._options.ssl.ca || []) : []
    }
  }

  private async startListening() {
    this._options.debug && this._options.debug("info", "Going to listen", this._portHttps, this._portHttp);
    this.emitEvent("beforeListen");

    const portHttp = this._portHttp;
    const portHttps = this._portHttps;

    // If we want to also listen to a HTTP fallback port and don't use greenlock(the library already has a
    // fallback mechanism built in) create a second express server that simply redirects everything to HTTPS.
    let fallback: Promise<number | null>;
    if (portHttp && this._options.ssl !== "letsEncrypt" && this._options.ssl !== "none") {
      const fallbackApp = express();
      fallbackApp.get("*", (req, res) => res.redirect("https://" + req.headers.host + ":" + this._portHttps + req.url));
      fallback = new Promise((resolve, reject) => {
        let fallbackServer;
        if (this._host) {
          fallbackServer = fallbackApp.listen(portHttp, this._host, (err: any) => err ? reject(err) : resolve(portHttp));
        } else {
          fallbackServer = fallbackApp.listen(portHttp, (err: any) => err ? reject(err) : resolve(portHttp));
        }
        this._server.push(enableDestroy(fallbackServer))
      });
    } else {
      fallback = Promise.resolve(portHttp);
    }

    // Start the HTTPS server. We either use greenlock for Let's Encrypt issued certificates, or spdy for 
    // manually generated certificates.
    let main: Promise<number | null>;
    if (this._options.ssl === "letsEncrypt") {
      // Let's Encrypt issued certificates
      if (this._options.isDevelopment) {
        process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";
      }
      main = new Promise((resolve, reject) => {
        let server;
        if (this._host) {
          server = greenlock
            .create({ ...this.greenlockOptions(), app: this._app })
            .listen(this._host + ":" + this._portHttp, this._host + ":" + this._portHttps,
              (err: any) => err ? reject(err) : resolve(this._portHttps));
        } else {
          server = greenlock
            .create({ ...this.greenlockOptions(), app: this._app })
            .listen(this._portHttp, this._portHttps,
              (err: any) => err ? reject(err) : resolve(this._portHttps));
        }
        this._server.push(enableDestroy(server))
      });
    } else if (this._options.ssl === "none") {
      // No certificates, HTTP only.
      main = new Promise((resolve, reject) => {
        if (this._portHttp === null) {
          return reject(new Error("Cannot listen on HTTP without HTTP port."));
        }
        let httpServer;
        if (this._host) {
          httpServer = this._app.listen(this._portHttp, this._host, (err: any) => err ? reject(err) : resolve(this._portHttp));
        } else {
          httpServer = this._app.listen(this._portHttp, (err: any) => err ? reject(err) : resolve(this._portHttp))
        }
        this._server.push(enableDestroy(httpServer));
      });
    } else {
      // Own certificate, SPDY.
      if (this._options.ssl.allowUnsigned || this._options.isDevelopment) {
        process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";
      }
      main = new Promise((resolve, reject) => {
        let server;
        if (this._host) {
          server = spdy.createServer(this.spdyOptions(), this._app).listen(portHttps!, this._host, (err: any) => err ? reject(err) : resolve(portHttps));
        } else {
          server = spdy.createServer(this.spdyOptions(), this._app).listen(portHttps!, (err: any) => err ? reject(err) : resolve(portHttps));
        }
        this._server.push(enableDestroy(server));
      });
    }

    await Promise.all([main, fallback]).then(() => {
      this._options.debug && this._options.debug("info", "Listening", this._portHttps, this._portHttp);
      this.emitEvent("listen")
    });
    return this._server;
  }
}

export interface IOptions {
  name: string;
  domains: string[];
  cookieSecret: string;
  webmasterMail: string;

  isDevelopment: boolean;
  ssl: "none" | "letsEncrypt" | { key: string, cert: string, ca?: string[], allowUnsigned: boolean };
  letsEncryptCertDirectory: string;
  agreeGreenlockTos: boolean;

  staticFiles: { from: string, serve: string } | null;

  maxPayloadSize: number;
  debug: ((level: string, ...args: any[]) => void) | null;

};

const createDefaultOptions = ({
  debug, isDevelopment, agreeGreenlockTos,
  staticFiles, maxPayloadSize, name, ssl, webmasterMail, domains, letsEncryptCertDirectory, cookieSecret,
}: Partial<IOptions>): IOptions => {

  if (name === undefined) throw new Error("name is required.");
  if (domains === undefined) throw new Error("domains is required.");
  if (cookieSecret === undefined) throw new Error("cookieSecret is required.");
  if (webmasterMail === undefined) throw new Error("webmasterMail is required.");

  return {
    debug: debug === undefined ? defaultDebug : debug,
    isDevelopment: isDevelopment === undefined ? false : isDevelopment,
    agreeGreenlockTos: agreeGreenlockTos === undefined ? false : agreeGreenlockTos,

    staticFiles: staticFiles === undefined ? null : staticFiles,
    maxPayloadSize: maxPayloadSize === undefined ? 100 * 1024 : maxPayloadSize,
    ssl: ssl === undefined ? "letsEncrypt" : ssl,
    letsEncryptCertDirectory: letsEncryptCertDirectory === undefined ? "./.certs/" : letsEncryptCertDirectory,
    name, webmasterMail, domains, cookieSecret
  };
};

const defaultDebug = (level: string, ...args: any[]) => console.log("[" + level + "]", "🏰 ", ...args);

export default Server;
